posthog/hogql/database/schema/channel_type.py

```diff
@@ -0,0 +1,94 @@
+# Create a virtual field that uses GA's channel grouping logic to group events into acquisition channels.
+# The source for this logic is:
+# UA: https://support.google.com/analytics/answer/3297892?hl=en
+# GA4: https://support.google.com/analytics/answer/9756891?hl=en
+
+# I'm not fully convinced that this approach will work on its own, as GA4 will have a lot more information on paid ads
+# than what we will have access to. We'll need to get this live and see what it looks like on Posthog data.
+from posthog.hogql.database.models import ExpressionField
+from posthog.hogql.parser import parse_expr
+
+
+def create_initial_domain_type(name: str):
+    return ExpressionField(
+        name=name,
+        expr=parse_expr(
+            """
+if(
+    properties.$initial_referring_domain = '$direct',
+    '$direct',
+    dictGetOrNull(
+        'ga4_channel_definition_dict',
+        'type',
+        cutToFirstSignificantSubdomain(coalesce(properties.$initial_referring_domain, ''))
+    )
+)
+"""
+        ),
+    )
+
+
+def create_initial_channel_type(name: str):
+    return ExpressionField(
+        name=name,
+        expr=parse_expr(
+            """
+multiIf(
+    match(properties.$initial_utm_campaign, 'cross-network'),
+    'Cross Network',
+
+    match(properties.$initial_utm_medium, '^(.*cp.*|ppc|retargeting|paid.*)$'),
+    CASE dictGetOrNull('ga4_channel_definition_dict', 'type', cutToFirstSignificantSubdomain(coalesce(properties.$initial_referring_domain, '')))
+        WHEN 'Shopping' THEN 'Paid Shopping'
+        WHEN 'Search' THEN 'Paid Search'
+        WHEN 'Video' THEN 'Paid Video'
+        WHEN 'Social' THEN 'Paid Social'
+        ELSE multiIf(
+            match(properties.$initial_utm_campaign, '^(.*(([^a-df-z]|^)shop|shopping).*)$'),
+            'Paid Shopping',
+            properties.$initial_utm_medium IN
+                ('display', 'banner', 'expandable', 'interstitial', 'cpm'),
+            'Display',
+            'Paid Other'
+            )
+    END,
+
+    properties.$initial_referring_domain = '$direct' AND (properties.$initial_utm_medium IS NULL OR properties.$initial_utm_medium = ''),
+    'Direct',
+
+    CASE dictGetOrNull('ga4_channel_definition_dict', 'type', cutToFirstSignificantSubdomain(coalesce(properties.$initial_referring_domain, '')))
+        WHEN 'Shopping' THEN 'Organic Shopping'
+        WHEN 'Search' THEN 'Organic Search'
+        WHEN 'Video' THEN 'Organic Video'
+        WHEN 'Social' THEN 'Organic Social'
```

Comment:
Can these "Organic"-s just be in the dict directly?

Reply:
The dict says whether it's Search/Shopping/Social/etc, and the presence or lack of other utm tags (e.g. utm_medium=ppc) tells us whether it's organic or paid.